src/bip322.rs

```diff
@@ -0,0 +1,337 @@
+        BIP322Error::ValidationError(e)
+    }
+}
+
+/// Bip322 Signatures
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Bip322Signature {
+    /// Legacy style. Only applicable for P2PKH message_challenge
+    Legacy(Signature, PublicKey),
+
+    /// Simple witness structure
+    Simple(Vec<Vec<u8>>),
+
+    /// Full `to_sign` transaction structure
+    Full(Transaction),
+}
+
+/// BIP322 validator structure
+/// A standard for interoperable signed messages based on the Bitcoin Script format,
+/// either for proving fund availability, or committing to a message as the intended
+/// recipient of funds sent to the invoice address.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Bip322<T: MiniscriptKey + ToPublicKey> {
+    /// Message to be signed
+    message: Vec<u8>,
+
+    /// Signature to verify the message
+    /// Optional value is used here because a validator structure can be
+    /// created without a BIP322Signature. Such structure can only produce
+    /// to_spend (or empty to_sign) transaction, but cannot validate them.
+    signature: Option<Bip322Signature>,
+
+    /// script_pubkey to define the challenge script inside to_spend transaction
+    /// here we take in descriptors to derive the resulting script_pubkey
+    message_challenge: Descriptor<T>,
+}
+
+impl<T: MiniscriptKey + ToPublicKey> Bip322<T> {
+    /// Create a new BIP322 validator
+    pub fn new(msg: &[u8], sig: Option<Bip322Signature>, addr: Descriptor<T>) -> Self {
+        Bip322 {
+            message: msg.to_vec(),
+            signature: sig,
+            message_challenge: addr,
+        }
+    }
+
+    /// Insert Signature inside BIP322 structure
+    pub fn insert_sig(&mut self, sig: Bip322Signature) {
+        self.signature = Some(sig)
+    }
+
+    /// create the to_spend transaction
+    pub fn to_spend(&self) -> Transaction {
+        // create default input and output
+        let mut vin = TxIn::default();
+        let mut vout = TxOut::default();
+
+        // calculate the message tagged hash
+        let msg_hash = MessageHash::hash(&self.message[..]).into_inner();
+
+        // mutate the input with appropriate script_sig and sequence
+        vin.script_sig = Builder::new()
+            .push_int(0)
+            .push_slice(&msg_hash[..])
+            .into_script();
+        vin.sequence = 0;
+
+        // mutate the value and script_pubkey as appropriate
+        vout.value = 0;
+        vout.script_pubkey = self.message_challenge.script_pubkey();
+
+        // create and return final transaction
+        Transaction {
+            version: 0,
+            lock_time: 0,
+            input: vec![vin],
+            output: vec![vout],
+        }
+    }
+
+    /// Create to_sign transaction
+    /// This will create a transaction structure with empty signature and witness field
+    /// its up to the user of the library to fill the Tx with appropriate signature and witness  
+    pub fn to_sign(&self) -> Transaction {
```

Comment:
This should take a satisfier as an argument and should call satisfy to fill the witness and script sig.

Reply:
This should also take in the `age` and `height` as arguments instead of hard coded zeeros

Reply:
Changing `to_sign` to `empty_to_sign`. 

The purpose of this function is to create an empty `to_sign` transaction template, which then can be filled with the correct signature depending upon the `BIP322Signature` types. Which is currently happening inside the `validate()` function here. 

https://github.com/rust-bitcoin/rust-miniscript/blob/c740598c5eedfca6dcba82462f915150b99070c3/src/bip322.rs#L180-L211

Here we are mutating the empty `to_sign` transaction with corresponding signatures. Which can be done with a satisfier too. But because a `Bip322Signature::Full` type signature will contain the entire `to_sign` transaction itself (which I think will be most of the cases, the other two cases are rather simplistic and more specific so can be hardcoded easily), using a separate satisfier to create the same `to_sign` feels like more roundabout than necessary.

`Bip322Signer` probably needs to use a satisfier in some way, but we can flesh that detail out later.

For `age` and `height`, I think it's better to provide these values in the validator structure itself. Though I am not sure if I have applied these values correctly in the `empty_to_spend` creation, please check.